lib/active_merchant/billing/gateways/trans_first_transaction_express.rb

```diff
@@ -0,0 +1,597 @@
+        "R1" => "The transaction was declined or returned, because the cardholder requested that payment of all recurring or installment payment transactions for a specific merchant account be stopped/ Reserved for client-specific use (declined)",
+        "Q1" => "Card Authentication failed/ Reserved for client-specific use (declined)",
+        "XA" => "Forward to Issuer/ Reserved for client-specific use (declined)",
+        "XD" => "Forward to Issuer/ Reserved for client-specific use (declined)",
+      }
+
+      EXTENDED_RESPONSE_MESSAGES = {
+        "B40K" => "Declined Post – Credit linked to unextracted settle transaction"
+      }
+
+      TRANSACTION_CODES = {
+        authorize: 0,
+        void_authorization: 2,
+
+        purchase: 1,
+        capture: 3,
+        void_purchase: 6,
+        void_capture: 6,
+
+        refund: 4,
+        blind_credit: 5,
+        void_refund: 13,
+        void_credit: 13,
+
+        verify: 9,
+
+        wallet_sale: 14,
+      }
+
+      def initialize(options={})
+        requires!(options, :gateway_id, :reg_key)
+        super
+      end
+
+      def purchase(amount, payment_method, options={})
+        if credit_card? payment_method
+          request = build_xml_transaction_request do |doc|
+            add_transaction_code(doc, :purchase)
+            add_payment_method(doc, payment_method)
+            add_contact(doc, payment_method.name, options)
+            add_amount(doc, amount)
+            add_order_number(doc, options)
+          end
+        else
+          request = build_xml_transaction_request do |doc|
+            add_transaction_code(doc, :wallet_sale)
+            add_amount(doc, amount)
+            add_wallet_id(doc, payment_method)
+          end
+        end
+
+        commit(request)
+      end
+
+      def authorize(amount, payment_method, options={})
+        if credit_card? payment_method
+          request = build_xml_transaction_request do |doc|
+            add_transaction_code(doc, :authorize)
+            add_payment_method(doc, payment_method)
+            add_contact(doc, payment_method.name, options)
+            add_amount(doc, amount)
+          end
+        else
+          request = build_xml_transaction_request do |doc|
+            add_transaction_code(doc, :authorize)
+            add_amount(doc, amount)
+            add_wallet_id(doc, payment_method)
+          end
+        end
+
+        commit(request)
+      end
+
+      def capture(amount, authorization, options={})
+        request = build_xml_transaction_request do |doc|
+          add_transaction_code(doc, :capture)
+          add_amount(doc, amount)
+          add_original_transaction_data(doc, authorization)
+        end
+
+        commit(request)
+      end
+
+      # Transaction Express has three types of possible void
+      # - void_authorization
+      # - void_purchase (alias: void_capture)
+      # - void_refund (alias: void_credit)
+      def void(authorization, options={})
```

Comment:
This is tricky one but, instead of having the user pass in what type of void it is, what we can do is return an `authorization` from the original tx that has the tx type embedded in it. Something that looks like `purchase:1298shad698a`.

Then, when an authorization is passed to this void method, we know what type the original tx was and can self-contain the logic that determines what void type to invoke here.

If you search AM for `split_authorization` you'll see a lot of other instances where this is required.